Add new enterprise onboarding guide (#13728)
pulumi/docs: content/docs/pulumi-cloud/get-started/onboarding-guide.md

---
title: Setting Up Pulumi for Your Organization
meta_desc: Learn how to onboard your entire team to Pulumi, and build out your infrastructure platform, with built-in security, best practices, and compliance.
menu:
    cloud:
        name: Onboarding guide
        parent: pulumi-cloud-get-started
        weight: 1
        identifier: pulumi-onboarding-guide
---

In this guide, you’ll find everything you need to know about using Pulumi within your organization, from setting things up to recommended usage patterns and practices.

## In this guide

* Setup and start using Pulumi Cloud, whether SaaS or self-hosted
* Set and achieve onboarding goals for your team’s first 30 days to three months
* Understand and adopt recommended practices for Pulumi’s suite of products
* Work through (and prepare for) common first-time user challenges

> ## Pro tips

Reviewer: Might drop this into a notes tip block:

{{% notes type="tip" %}}
**Pro tips:**
To help everyone quickly adopt Pulumi in your organization, identify a group of “Pulumi champions” across your organization who will encourage other team members to get started right away. The faster your team uses Pulumi, the faster you can ship.
Check out the Pulumi Glossary for specific terms we use across our sites and documentation.
{{% /notes %}}

* To help everyone quickly adopt Pulumi in your organization, identify a group of “Pulumi champions” across your organization who will encourage other team members to get started right away. The faster your team uses Pulumi, the faster you can ship.
* Check out the [Pulumi Glossary](/docs/iac/concepts/glossary/) for specific terms we use across our sites and documentation.

## Part one: Setting up Pulumi Cloud

Setting up your Pulumi Cloud account will lay the foundation for onboarding your team and enabling collaboration. We’ll walk through the basics of getting your teams ready to build.

### Choosing SaaS versus self-hosted

Pulumi Cloud is available in two flavors: 1\) a multi-tenanted, secure, and compliant software-as-a-service (SaaS), and 2\) a self-hosted edition. These are referred to as simply Pulumi Cloud and Self-Hosted Pulumi Cloud, respectively.

> The easiest option is to choose SaaS, as things like high availability, disaster recovery, geo-replication, and more, are available out of the box. It is secure and compliant and its unique architecture allows you precise control over security, as described in the [Pulumi Cloud Security Whitepaper](https://www.pulumi.com/security/pulumi-cloud-security-whitepaper.pdf). To choose this option, simply log into [pulumi.com](http://pulumi.com).

Reviewer: ```suggestion
The easiest option is to choose SaaS, as things like high availability, disaster recovery, geo-replication, and more are available out of the box. It is secure and compliant and its unique architecture allows you precise control over security, as described in the [Pulumi Cloud Security Whitepaper](https://www.pulumi.com/security/pulumi-cloud-security-whitepaper.pdf). To choose this option, simply log into [pulumi.com](http://pulumi.com).
```

> For companies who require more control over the Pulumi Cloud hosting environment, such as would be the case for airgapped or or other highly regulated environments like FedRAMP, the self-hosted edition can be hosted anywhere: on-premises, in your cloud account, on a virtual machine, in a Kubernetes cluster, and so on. Learn more about [self-hosting Pulumi Cloud here](https://www.pulumi.com/product/self-hosted/).

Reviewer: ```suggestion
For companies who require more control over the Pulumi Cloud hosting environment, such as would be the case for air-gapped or other highly regulated environments like FedRAMP, the self-hosted edition can be hosted anywhere: on-premises, in your cloud account, on a virtual machine, in a Kubernetes cluster, and so on. Learn more about [self-hosting Pulumi Cloud here](https://www.pulumi.com/product/self-hosted/).
```

### Getting familiar with the management console

As soon as you are up and running, you will want to familiarize yourself with the Pulumi Cloud console. The console is accessible via the “Sign In” link in the upper right corner of [pulumi.com](http://pulumi.com) or by going to [app.pulumi.com](http://app.pulumi.com) directly. You’ll land on a Dashboard with useful content and links and the left navigation bar features product-specific tabs as well as Settings. From the console, you can manage your teams, see your stacks and resources, search for resources, and more. Your personal account is accessible from your avatar in the upper right corner and there is also a sparkle icon in the lower right corner which opens Pulumi Copilot, a smart AI assistant.

## Part two: Managing your teams

Pulumi Cloud makes it easy to collaborate on your cloud projects faster and with more confidence. Typically the infrastructure platform group will own the configuration of Pulumi and onboarding developers and security experts to facilitate building secure cloud applications and infrastructure. The onboarding process begins with configuring your organization.

### Organizing people for collaboration

> Your Pulumi Cloud hierarchy is made up of individual **users**, **organizations**, and **teams**. Each plays a critical role in how your team collaborates and enforcing security and best practices.

Reviewer: ```suggestion
Your Pulumi Cloud hierarchy is made up of individual **users**, **organizations**, and **teams**. Each plays a critical role in how your team collaborates and enforces security and best practices.
```

Start by creating [your organization here](https://app.pulumi.com/?create-organization=1). Organizations and their capabilities are a paid feature of Pulumi Cloud, so you will notice that creating an organization starts a free trial. Pulumi has a generous free tier when the trial expires and you can always [contact us](https://pulumi.com/contact) if you need more time.

### Configuring Single Sign-On (SSO)

Many organizations prefer to use Single Sign-On (SSO) to leverage an identity provider such as Azure Active Directory, Google Workspace, Okta, or any other SAML 2.0 compliant provider. To learn about configuring SSO for your organization, [read Pulumi Cloud SAML(SSO)](https://www.pulumi.com/docs/pulumi-cloud/access-management/saml/). If you’d like to continue using email, GitHub, GitLab, or Atlassian identity, you can skip this step.

### Inviting your teammates

> As soon as you’ve completed these set up steps, you are ready to invite your team\! If you are using SSO with System for Cross-domain Identity Management (SCIM) enabled, onboarding and offboarding happen automatically. Otherwise, refer to [Inviting members to an organization](https://www.pulumi.com/docs/pulumi-cloud/admin/organizations/#inviting-members-to-an-organization).

Reviewer: ```suggestion
As soon as you’ve completed these set up steps, you are ready to invite your team\! If you have SSO with System for Cross-domain Identity Management (SCIM) enabled, onboarding and offboarding happen automatically. Otherwise, refer to [Inviting members to an organization](https://www.pulumi.com/docs/pulumi-cloud/admin/organizations/#inviting-members-to-an-organization).
```

### Setting up cloud access methods

Pulumi supports hundreds of cloud providers, but most organizations will end up using at least AWS, Azure, Google Cloud, and/or Kubernetes. Other supported providers include SaaS infrastructure products, such as Cloudflare, DataDog, MongoDB, and Snowflake, as well as on-premises technologies such as VMWare vSphere. For a complete list of Pulumi providers, check out the [Pulumi Registry](https://pulumi.com/registry). The registry is your one- stop shop for provider documentation, including how to configure access to each of the clouds, as well as resource APIs.

Although each cloud offers flexible authentication unique to that cloud, Pulumi ESC offers rich OpenID Connect (OIDC) support for several popular providers, ensuring dynamic, short-lived credentials. This technique is the most secure and should be preferred for those providers that have support. Learn more about [dynamic login credentials here](https://www.pulumi.com/docs/esc/integrations/dynamic-login-credentials/).

> In the event your chosen cloud doesn’t have a Pulumi ESC OIDC provider, refer to the registry documentation. Each provider has an “Install & config” section on the left- hand navigation. See [AWS Installation & Configuration](https://www.pulumi.com/docs/esc/integrations/dynamic-login-credentials/) as an example. Pulumi generally uses native tools and techniques for authenticating so it is idiomatic and consistent with your other usage patterns.

Reviewer: ```suggestion
In the event your chosen cloud doesn’t have a Pulumi ESC OIDC provider, refer to the registry documentation. Each provider has an “Install & config” section on the left-hand navigation. See [AWS Installation & Configuration](https://www.pulumi.com/docs/esc/integrations/dynamic-login-credentials/) as an example. Pulumi generally uses native tools and techniques for authenticating so it is idiomatic and consistent with your other usage patterns.
```

### Deploying continuously

Although most teams will get started by running the Pulumi CLI and performing activities in the Pulumi Cloud console manually, at some stage they will want to automate certain workflows.

The most common workflow is deploying changes to your IaC configurations. The Pulumi CLI can be used to preview and deploy changes. But most teams will prefer to trigger deployments using a Git-based workflow using one of Pulumi’s many continuous integration and deployment (CI/CD) integrations. This approach generally delivers previews in your pull requests, facilitating code reviews and testing, and then performs updates upon merging. Once your organization is comfortable with this basic CI/CD workflow, Pulumi recommends adding drift detection to your stacks on a scheduled basis. (We’ll show you how to do this later on.)

> Commonly a team’s “official” stacks like staging and production will use CI/CD-based approach while individual developers still work on development stacks manually. Many teams already use CI/CD for application delivery and adopting it for infrastructure delivery helps to unify practices.

Reviewer: ```suggestion
Commonly, a team’s “official” stacks, like staging and production, will use a CI/CD-based approach while individual developers still work on development stacks manually. Many teams already use CI/CD for application delivery and adopting it for infrastructure delivery helps to unify practices.
```

These workloads should be automated with CI/CD pipelines and generally use as many of the best practices in this document as possible, even though at this stage you’ll be uncomfortably early on your journey to figuring out things like which components and templates to build.

### Stay focused on impact

This initial beachhead win will inform where you go from here. Rather than starting out by creating dozens of abstractly-useful and broad components, it is better to take a “workload-first” strategy, and inform very specific component requirements from real world applications.

> We also recommend resisting the temptation to conflate redesigning your projects to use new cloud architectures with the migration to your platform. This adds risk. Sometimes this is inevitable, especially if the shift to a platform coincides with a major initiative like moving from on-prem to the cloud, but avoid it if you can. If possible, get workloads onto Pulumi and your platform first – from there it is then easier to refactor and redesign projects in place.

Reviewer: ```suggestion
We also recommend resisting the temptation to conflate redesigning your projects to use new cloud architectures with the migration to your platform. This adds risk. Sometimes this is inevitable, especially if the shift to a platform coincides with a major initiative like moving from on-prem to the cloud, but avoid it if you can. If possible, get workloads onto Pulumi and your platform first. From there, it is easier to refactor and redesign projects.
```

### Measure your success

Put in place clear success metrics from the outset. Common metrics we see Pulumi customers improving dramatically include “Time from checking in code until it ships to production.” Related, the DevOps Research and Assessments (DORA) metric of deployment frequency and related metrics are useful industry standard measurements. It’s common to see improvements of 5-10X in these areas for Pulumi customers. Another metric is how long it takes for a developer to get a new cloud environment to work with, and by adopting some of the self-serve techniques in this guide, many Pulumi customers are able to reduce this from months to minutes.

## Part nine: Getting support

We aspire to make Pulumi as easy to use as possible, and hope that this guide helps you navigate your way around the Pulumi ecosystem. That said, we are here to help with anything you need, whether it’s advice on your architecture, help using a particular cloud resource, training for your team, or anything in between.

### Learning your way around examples

> One of the most challenging aspects of any IaC solution, Pulumi included, is learning how to configure the cloud resources you desire to use. Between 100s of clouds, and 100s or 1,000s of services in each of them, and dozens of configuration settings per service, there are endless ways to configure them. Pulumi supports containerized workloads in the cloud as well as serverless and on-premises VM-based workloads. This is very powerful–but can be daunting\!

Reviewer: ```suggestion
One of the most challenging aspects of any IaC solution, Pulumi included, is learning how to configure the cloud resources you desire to use. Between 100s of clouds, and 100s or 1,000s of services in each of them, and dozens of configuration settings per service, there are endless ways to configure them. Pulumi supports containerized workloads in the cloud as well as serverless and on-premises VM-based workloads. This is very powerful–but can be daunting!
```
